sources/web/datalab/polymer/components/file-browser/file-browser.ts

```diff
@@ -336,9 +371,17 @@ class FileBrowserElement extends Polymer.Element implements DatalabPageElement {
         this._pathHistory.slice(1, this._pathHistoryIndex + 1).map((p) => p.name);
 
     this.currentFile = this._pathHistory[this._pathHistoryIndex];
+    this._setFileParamToCurrentFile();
     this._fetchFileList();
   }
 
+  _setFileParamToCurrentFile() {
+    // Only update the location if we are currently visible
+    if (this.currentFile && this.offsetParent) {
+      this.set('queryParams.file', this.currentFile.id.toQueryString());
```

Comment:
If we do something like `window.history.pushState` here, doesn't this already move all the history logic to the browser?

Reply:
.. And then the back/forward buttons of course will need to do `window.history.{backward|forward}()`.

Reply:
Updating queryParams this way actually does add to the browser history. I added some code to observe changes to queryParams and update the page, which makes the browser's backwards and forwards buttons work properly - and it doesn't reload the page. There is some kind of problem when doing this with drive, though: it is sending a query to the drive API, but not updating the screen. I spent a bit of time trying to track that down, with no success yet. I don't know if it's a problem in drive-file-manager, file-browser, or how they connect.

Reply:
Added some code that makes browser history work on drive also. So now the only oddity I am aware of is that the breadcrumbs in drive don't get set up on direct entry of the file parameter, and thus are not preserved browser forward/backwards.